### PR TITLE
Restore signals on teardown

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Restore the default signal handlers on teardown if Lightning configured their own.

Currently, the Trainer / SignalConnector configures signals when running on SLURM or when fault-tolerant training is enabled. These signals are configured in global namespace, and therefore remain intact after Trainer exits. This may not be a problem normally, but one could run into hard to debug issues when running PL in unit tests. This PR makes sure the signals are properly restored.

Found while working on #10601 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @awaelchli @carmocca @justusschock @ninginthecloud @kaushikb11

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/signal_connector.py [hunk 1]
```diff
@@ -27,11 +27,16 @@ def __call__(self, signum: Signals, frame: FrameType) -> None:
...
 
 class SignalConnector:
-    def __init__(self, trainer: "pl.Trainer"):
+    def __init__(self, trainer: "pl.Trainer") -> None:
         self.trainer = trainer
         self.trainer._terminate_gracefully = False
+        self._original_handlers: Dict[int, Any] = {}
 
     def register_signal_handlers(self) -> None:
+        self._original_handlers = {signal.SIGTERM: signal.getsignal(signal.SIGTERM)}
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/signal_connector.py:36]: Can we move this within a function ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/connectors/signal_connector.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/signal_connector.py]: done!

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/signal_connector.py [hunk 2]
```diff
@@ -101,12 +101,33 @@ def teardown(self) -> None:
...
+            return signal.valid_signals()
+        elif _IS_WINDOWS:
+            return {
+                signal.SIGABRT,
+                signal.SIGFPE,
+                signal.SIGILL,
+                signal.SIGINT,
+                signal.SIGSEGV,
+                signal.SIGTERM,
+                signal.SIGBREAK,
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/signal_connector.py:131]: How did you find which signals go into this list? adding a comment with the reference would be nice.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/signal_connector.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/signal_connector.py:131]: from the Python docs :) added reference

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**